src/Sylius/Bundle/CoreBundle/Application/SyliusPluginTrait.php

```diff
@@ -0,0 +1,109 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sylius\Bundle\CoreBundle\Application;
+
+use Symfony\Component\DependencyInjection\Container;
+use Symfony\Component\DependencyInjection\Extension\ExtensionInterface;
+
+/**
+ * @mixin \Symfony\Component\HttpKernel\Bundle\Bundle
+ * @see \Symfony\Component\HttpKernel\Bundle\Bundle
+ *
+ * Provides a common logic for Sylius Plugins.
+ * Each of a plugins should be created with Plugin instead of Bundle suffix for the root class.
+ *
+ * @author Łukasz Chruściel <lukasz.chrusciel@lakion.com>
+ */
+trait SyliusPluginTrait
+{
+    /**
+     * Returns the plugin's container extension.
+     *
+     * @return ExtensionInterface|null The container extension
+     *
+     * @throws \LogicException
+     */
+    public function getContainerExtension()
+    {
+        if (null === $this->getExtension()) {
+            $extension = $this->createContainerExtension();
+
+            if (null !== $extension) {
+                if (!$extension instanceof ExtensionInterface) {
+                    throw new \LogicException(sprintf('Extension %s must implement %s.', get_class($extension), ExtensionInterface::class));
+                }
+
+                // check naming convention for Sylius Plugins
+                $basename = preg_replace('/Plugin$/', '', $this->getName());
+                $expectedAlias = Container::underscore($basename);
+
+                if ($expectedAlias != $extension->getAlias()) {
+                    throw new \LogicException(sprintf(
+                        'Users will expect the alias of the default extension of a plugin to be the underscored version of the plugin name ("%s"). You can override "Bundle::getContainerExtension()" if you want to use "%s" or another alias.',
+                        $expectedAlias, $extension->getAlias()
+                    ));
+                }
+
+                $this->setExtension($extension);
+            } else {
+                $this->setExtension(false);
+            }
+        }
+
+        if ($this->getExtension()) {
+            return $this->getExtension();
+        }
+    }
+
+    /**
+     * Creates the bundle's container extension.
+     *
+     * @return ExtensionInterface|null
+     */
+    abstract protected function createContainerExtension();
+
+    /**
+     * Returns the bundle name (the class short name).
+     *
+     * @return string The Bundle name
+     */
+    abstract protected function getName();
+
+    /**
+     * Gets the Bundle namespace.
+     *
+     * @return string The Bundle namespace
+     */
+    abstract protected function getNamespace();
+
+    /**
+     * @return ExtensionInterface|bool
+     */
+    abstract protected function getExtension();
```

Comment:
There should be no setters / getters for extension like this. Outside code should always use `getContainerExtension()` and these setters / getters should be replaced with a private property.